kani-compiler/src/kani_middle/transform/check_uninit/mod.rs

```diff
@@ -0,0 +1,433 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+//! Implement a transformation pass that instruments the code to detect possible UB due to
+//! the accesses to uninitialized memory.
+
+use crate::args::ExtraChecks;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::body::{
+    CheckType, InsertPosition, MutableBody, SourceInstruction,
+};
+use crate::kani_middle::transform::{TransformPass, TransformationType};
+use crate::kani_queries::QueryDb;
+use lazy_static::lazy_static;
+use rustc_middle::ty::TyCtxt;
+use rustc_smir::rustc_internal;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{AggregateKind, Body, ConstOperand, Mutability, Operand, Place, Rvalue};
+use stable_mir::ty::{
+    FnDef, GenericArgKind, GenericArgs, MirConst, RigidTy, Ty, TyConst, TyKind, UintTy,
+};
+use stable_mir::CrateDef;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::sync::Mutex;
+use tracing::{debug, trace};
+
+mod ty_layout;
+mod uninit_visitor;
+
+pub use ty_layout::{PointeeInfo, PointeeLayout, TypeLayout};
+use uninit_visitor::{CheckUninitVisitor, InitRelevantInstruction, SourceOp};
+
+const SKIPPED_DIAGNOSTIC_ITEMS: &[&str] = &["KaniMemInitSMGetInner", "KaniMemInitSMSetInner"];
+
+/// Retrieve a function definition by diagnostic string, caching the result.
+pub fn get_kani_sm_function(tcx: TyCtxt, diagnostic: &'static str) -> FnDef {
+    lazy_static! {
+        static ref KANI_SM_FUNCTIONS: Mutex<HashMap<&'static str, FnDef>> =
```

Comment:
Changed it. As a part of that, I had to make the `transform` function of each pass accept `self` as mutable, since the cache is updated as the pass progresses

Reply:
I was thinking that you could cache all functions you need as part of the pass constructor. But that should be fine too.